zebra-state/src/service/non_finalized_state/chain.rs

```diff
@@ -521,11 +532,20 @@ impl Chain {
         let anchor = tree.root();
         trace!(?height, ?anchor, "adding sprout tree");
 
-        assert_eq!(
-            self.sprout_trees_by_height.insert(height, tree.clone()),
-            None,
-            "incorrect overwrite of sprout tree: trees must be reverted then inserted",
-        );
+        // Don't add a new tree unless it differs from the previous one or there's no previous tree.
+        if height.is_min()
+            || self
+                .sprout_tree(height.previous().into())
+                .map_or(true, |prev_tree| prev_tree != tree)
```

Comment:
I don't think these extra steps are necessary, hashing the roots to check if the sapling/orchard trees are the same looks expensive, and it would only clone the `Arc` for the previous tree unless it's been modified:

These trees come from `X_trees_by_height.last_key_value().unwrap().clone()`: https://github.com/ZcashFoundation/zebra/blob/release-v1.1.0/zebra-chain/src/parallel/tree.rs#L104-L124 
https://github.com/ZcashFoundation/zebra/pull/7259#discussion_r1268735828

Reply:
The roots for each tree are already calculated and cached, because we use them to verify the history tree. (Or index the sprout tree roots.)

Reply:
So I don't think there's any CPU impact here. However, it is faster to compare the inner frontier if the roots aren't cached. We could fix that in the `PartialEq` impl.

Reply:
> The roots for each tree are already calculated and cached

I meant hashing the roots. It's also a second lookup when reading/writing or a bit more searching when reading.

> So I don't think there's any CPU impact here.

I don't think there should be a significant impact either.

Reply:
Where do you think the roots get hashed?

Unless I miss something, the roots are the resulting hashes of the trees, and the trees cache them.

Reply:
We basically compare `[u8; 32]` https://github.com/zcashfoundation/zebra/blob/9ebd56092bcdfc1a09062e15a0574c94af37f389/zebra-chain/src/sprout/tree.rs#L100.

Reply:
> I don't think these extra steps are necessary

I couldn't come up with anything better than comparing the previous tree. Is there anything else we can do?

Reply:
> However, it is faster to compare the inner frontier if the roots aren't cached. We could fix that in the PartialEq impl.

Let's do that. I'll open an issue.

Reply:
> We basically compare `[u8; 32]` https://github.com/zcashfoundation/zebra/blob/9ebd56092bcdfc1a09062e15a0574c94af37f389/zebra-chain/src/sprout/tree.rs#L100.

Ah, sorry, it's sapling/orchard, not sprout:

https://github.com/ZcashFoundation/zebra/blob/dedup-tree-insertion-fin-state/zebra-chain/src/sapling/tree.rs#L109

https://github.com/ZcashFoundation/zebra/blob/dedup-tree-insertion-fin-state/zebra-chain/src/orchard/tree.rs#L116

Reply:
> I couldn't come up with anything better than comparing the tree being stored to the previous tree. Is there anything else we can do?

It could always clone and insert the value, even when it's the same as the previous tree. The `BTreeMap`s for `x_trees_by_height` would use a bit more memory, but the value would be an `Arc` referencing the previous tree.

Reply:
A check for `Arc` equality works if they are equal, but if the `Arc`s aren't equal, we still need to compare either the cached root or entire frontier.

This logic is already implemented in `Arc::eq()`, so we don't need to do anything extra to take advantage of it:
https://doc.rust-lang.org/std/sync/struct.Arc.html#impl-PartialEq%3CArc%3CT%3E%3E-for-Arc%3CT%3E

We can compare `Arc<Tree>` or `Arc<Root>` and it will just work.

Reply:
> we still need to compare either the cached root or entire frontier.

I think it should be okay to check if it's stored in the same allocation because it should always be equal when there are no new note commitments applied to the previous tree or unequal otherwise.

Reply:
It's possible to create a fresh clone of a note commitment tree from the finalized state at any time. If that tree has the same notes as an existing tree, it must compare equal, but it won't be in the same `Arc`.

(It's also possible to clone the inner value of an `Arc` into a new allocation using the `Arc` methods.)

Reply:
Thanks for this discussion.